### ISSUE TITLE
openssl 3.0.0 compatible aws-sdk-cpp version

### ISSUE THREAD

**Issue by satya-rajesh** [aws/aws-sdk-cpp#1582]
While compiling observed that aws-sdk-cpp is not compatible with openssl 3.0.0.

aws-sdk-cpp is using deprecated APIs which are not present in openssl 3.0.0 (alpha).

Build status:
Failed due to HMAC_* APIs deprecated

Please provide aws-sdk-cpp release compatible to openssl 3.0.0

**Comment by jmklix** [comment]
We don't currently support the latest version of openssl 3.0.0, but we are interested to hear you use case and what features you want to use with it.

**Comment by satya-rajesh** [comment]
> We don't currently support the latest version of openssl 3.0.0, but we are interested to hear you use case and what features you want to use with it.

openssl 3.0 has deprecated several APIs to it's previous version:
https://www.openssl.org/news/changelog.html#openssl-30

We tried to build aws-sdk-cpp with openssl 3.0, since aws-sdk-cpp still has deprecated API's, it failed to build.

**Comment by thompsow** [comment]
I would like an update on this issue. My interest in this is relative to FIPS support in OpenSSL in our project which uses OpenSSL 3.0.0. Our use case is for AWS-SDK-CPP is for interop with our on-prem solution and cloud (AWS) solution.

**Comment by olcryx** [comment]
> I would like an update on this issue. My interest in this is relative to FIPS support in OpenSSL in our project which uses OpenSSL 3.0.0. Our use case is for AWS-SDK-CPP is for interop with our on-prem solution and cloud (AWS) solution.

Second this. OpenSSL 1.1.x does not support FIPS configuration, and OpenSSL 1.0.x is EOL. OpenSSL 3.x _does_ support FIPS configuration, so it's pretty much our only option going forward.

**Comment by breuner** [comment]
@jmklix any update on this available? seems like this makes the aws sdk cpp incompatible with the soon to be released ubuntu 22.04, because [ubuntu 22.04 comes with openssl 3.0](https://linuxconfig.org/ubuntu-22-04-features-and-release-date) - so i think this is more than just a feature request now.
```
cmake . -DBUILD_ONLY="s3;transfer" -DBUILD_SHARED_LIBS=OFF -DCPP_STANDARD=14 \
        -DAUTORUN_UNIT_TESTS=OFF -DENABLE_TESTING=OFF \
        -DCMAKE_BUILD_TYPE=Release -DBYO_CRYPTO=ON \
        "-DCMAKE_INSTALL_PREFIX=$INSTALL_DIR" && \
    make -j $(nproc) install
[...}
[ 98%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/ub_core.cpp.o
In file included from /root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/ub_core.cpp:28:
/root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp: In constructor 'Aws::Utils::Crypto::HMACRAIIGuard::HMACRAIIGuard()':
/root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp:416:41: error: 'HMAC_CTX* HMAC_CTX_new()' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
  416 |                     m_ctx = HMAC_CTX_new();
      |                             ~~~~~~~~~~~~^~
In file included from /root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h:14,
                 from /root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp:9,
                 from /root/elbencho/external/aws-sdk-cpp/aws-cpp-sdk-core/ub_core.cpp:28:
/usr/include/openssl/hmac.h:33:33: note: declared here
   33 | OSSL_DEPRECATEDIN_3_0 HMAC_CTX *HMAC_CTX_new(void);
      |                                 ^~~~~~~~~~~~
[...]
cc1plus: all warnings being treated as errors
make[3]: Leaving directory '/root/elbencho/external/aws-sdk-cpp'
make[3]: *** [aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/build.make:932: aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/ub_core.cpp.o] Error 1
make[2]: Leaving directory '/root/elbencho/external/aws-sdk-cpp'
make[2]: *** [CMakeFiles/Makefile2:758: aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/all] Error 2
make[1]: Leaving directory '/root/elbencho/external/aws-sdk-cpp'
make[1]: *** [Makefile:136: all] Error 2
```

**Comment by tobodner** [comment]
While the above mentioned APIs are deprecated, they in fact _are_ still available. However, the deprecation warning may be treated as an error by your compiler. Adding the flag `Wno-error=deprecated-declarations` may do the job for you for now.

**Comment by dav1d-wright** [comment]
I have run into the same issue after upgrading to Ubuntu 22.04. I was able to work around this by specifying` -Wno-error=deprecated-declarations` or `-DUSE_OPENSSL=OFF`, however neither of this is a satisfying solution, as one is a temporary workaround which will fail as soon as the APIs are removed from openssl, and the other replaces openssl with an experimental feature, according to the AWS_SDK CMakeLists.txt

**Comment by rhys-67bricks** [comment]
I spent a lot of time trying out different options. Eventually this worked for me:
`cmake -DCMAKE_CXX_FLAGS="-Wno-error=deprecated-declarations" -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DENABLE_TESTING=OFF ..`

**Comment by RuurdBeerstra** [comment]
I also ran into this problem, after we recently upgraded our software base to use OpenSSL 3.0 (as advised by OpenSSL). I fixed  it by patching the file aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp which contains the calls to the deprecated functions (not that hard). A simple change to the new functions with the same signature make the problems go away. I do use the compatibility flags for OpenSSL by passing this to cmake:
-DCMAKE_CXX_FLAGS="-DOPENSSL_API_COMPAT=0x10100000L"

That allowed me to build AWS using OpenSSL 3.0 without further issue. I'll attach the patch in case it is useful for someone else.
[CryptoImpl.patch.txt](https://github.com/aws/aws-sdk-cpp/files/8900557/CryptoImpl.patch.txt)
